app/src/main/java/io/appium/uiautomator2/core/AccessibilityNodeInfoHelpers.java

```diff
@@ -82,43 +84,62 @@ public static String getText(@Nullable AccessibilityNodeInfo nodeInfo, boolean r
+            node.getBoundsInScreen(rect);
+            return rect;
+        }
+
+        UiDevice uiDevice = getUiDevice();
+        Rect screenRect = new Rect(0, 0, uiDevice.getDisplayWidth(), uiDevice.getDisplayHeight());
+        return getBounds(node, screenRect, 0);
     }
 
     /**
      * Returns the node's bounds clipped to the size of the display, limited by the MAX_DEPTH
+     * The implementation is borrowed from `getVisibleBounds` method of `UiObject2` class
      *
-     * @return null if node is null, else a Rect containing visible bounds
+     * @return Empty rect if node is null, else a Rect containing visible bounds
      */
     @SuppressLint("CheckResult")
-    private static Rect getVisibleBounds(@Nullable AccessibilityNodeInfo node, int depth) {
+    private static Rect getBounds(@Nullable AccessibilityNodeInfo node, Rect displayRect, int depth) {
+        Rect ret = new Rect();
         if (node == null) {
-            return null;
+            return ret;
         }
 
         // Get the object bounds in screen coordinates
-        Rect ret = new Rect();
         node.getBoundsInScreen(ret);
-        UiDevice uiDevice = getUiDevice();
 
         // Trim any portion of the bounds that are not on the screen
-        Rect screen = new Rect(0, 0, uiDevice.getDisplayWidth(), uiDevice.getDisplayHeight());
-        ret.intersect(screen);
+        ret.intersect(displayRect);
 
-        // Find the visible bounds of our first scrollable ancestor 
-        for (AccessibilityNodeInfo ancestor = node.getParent(); ancestor != null && ++depth < MAX_DEPTH; ancestor = ancestor.getParent()) {
-            // If this ancestor is scrollable
-            if (ancestor.isScrollable()) {
-                // Trim any portion of the bounds that are hidden by the non-visible portion of our
-                // ancestor
-                Rect ancestorRect = getVisibleBounds(ancestor, depth);
-                ret.intersect(ancestorRect);
-                break;
+        // Trim any portion of the bounds that are outside the window
+        Rect window = new Rect();
+        if (node.getWindow() != null) {
+            node.getWindow().getBoundsInScreen(window);
+            ret.intersect(window);
+        }
+
+        // Find the visible bounds of our first scrollable ancestor
+        AccessibilityNodeInfo ancestor;
+        int currentDepth = depth;
+        for (ancestor = node.getParent(); ancestor != null && ++currentDepth < MAX_DEPTH; ancestor = ancestor.getParent()) {
+            if (!ancestor.isScrollable()) {
+                continue;
             }
+            // If this ancestor is scrollable
+            // Trim any portion of the bounds that are hidden by the non-visible portion of our
+            // ancestor
+            Rect ancestorRect = getBounds(ancestor, displayRect, currentDepth);
+            ret.intersect(ancestorRect);
+            break;
```

Comment:
Codacy found an issue: [Avoid using a branching statement as the last in a loop.](https://app.codacy.com/manual/dpgraham/appium-uiautomator2-server/pullRequest?prid=5788633)